File: app/src/main/java/com/example/NaiveAAC/activities/Account/AccountActivity.java
package com.example.NaiveAAC.activities.Account;

import static com.example.NaiveAAC.activities.Settings.Utils.AdvancedSettingsDataImportExportHelper.findExternalStorageRoot;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.view.View;
import android.widget.EditText;

import androidx.core.app.ActivityCompat;
import androidx.fragment.app.FragmentManager;

import com.example.NaiveAAC.R;
import com.example.NaiveAAC.activities.Game.ChoiseOfGame.ChoiseOfGameActivity;
import com.example.NaiveAAC.activities.Game.GameParameters.GameParameters;
import com.example.NaiveAAC.activities.Game.Utils.ActionbarFragment;
import com.example.NaiveAAC.activities.Grammar.GrammaticalExceptions;
import com.example.NaiveAAC.activities.Grammar.ListsOfNames;
import com.example.NaiveAAC.activities.Graphics.Images;
import com.example.NaiveAAC.activities.Graphics.Videos;
import com.example.NaiveAAC.activities.Phrases.Phrases;
import com.example.NaiveAAC.activities.Settings.AccountFragment;
import com.example.NaiveAAC.activities.Settings.Utils.AccountActivityAbstractClass;
import com.example.NaiveAAC.activities.Settings.Utils.AdvancedSettingsDataImportExportHelper;
import com.example.NaiveAAC.activities.Settings.Utils.SettingsFragmentAbstractClass;
import com.example.NaiveAAC.activities.Stories.Stories;
import com.example.NaiveAAC.activities.WordPairs.WordPairs;
import com.example.NaiveAAC.activities.history.History;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;

import io.realm.Realm;
import io.realm.RealmResults;

/**
 * <h1>AccountActivity</h1>
 * <p><b>AccountActivity</b> create from assets the initial realm database , register the user
 * and move on to the welcome activity.</p>
 * the initial realm database consists of
 * </p>
 * 1) data from the Arasaac /pictograms/all API</p>
 * 2) data from https://github.com/ian-hamlin/verb-data a collection of verbs and conjugations </p>
 * 3) initial settings and content such as images, videos and others from assets</p>
 *
 * @version     1.1, 04/22/22
 * @see com.example.NaiveAAC.activities.Settings.Utils.AccountActivityAbstractClass
 * @see com.example.NaiveAAC.activities.Settings.Utils.SettingsFragmentAbstractClass
 */
public class AccountActivity extends AccountActivityAbstractClass implements
        SettingsFragmentAbstractClass.onFragmentEventListenerSettings         {
    //
    private static final String TAG = "VERBO";
    private static final String TAGPERMISSION = "Permission";
    //
    public static final String EXTRA_MESSAGE ="helloworldandroidMessage" ;
    //
    public FragmentManager fragmentManager;
    //
    private final String FOLDER_SIMSIM = "simsim";
    //
    public String textPersonName;
    //
    public File root;
    public String rootPath;
    //
    /**
     * configurations of account start screen.
     *
     * @param savedInstanceState Define potentially saved parameters due to configurations changes.
     * @see #setActivityResultLauncher
     * @see ActionbarFragment
     * @see AccountFragment
     * @see android.app.Activity#onCreate(Bundle)
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_account);
        //
        context = this;
        //
        setActivityResultLauncher();
        //
        if (savedInstanceState == null)
            {
            fragmentManager = getSupportFragmentManager();
            fragmentManager.beginTransaction()
                    .add(new ActionbarFragment(), getString(R.string.actionbar_fragment))
                    .add(R.id.settings_container, new AccountFragment(), getString(R.string.account_fragment))
                    .commit();
            }
        //
        if (isStoragePermissionGranted()) {
        //    Log.v(TAGPERMISSION,getString(R.string.permission_is_granted));
        }
        //
    }
    /**
     * Called when the user taps the save account button.
     * </p>
     * if the case creates creates the initial realm database.</p>
     *
     * @param view view of tapped button
     * @see #isStoragePermissionGranted
     * @see AdvancedSettingsDataImportExportHelper#findExternalStorageRoot()
     * @see #prepareTheSimsimDirectory
     * @see #registerPersonName(String)
     * @see AdvancedSettingsDataImportExportHelper#openFileInput(Context, String)
     * @see Images
     * @see Videos
     * @see GameParameters
     * @see GrammaticalExceptions
     * @see ListsOfNames
     * @see Phrases
     * @see Stories
     * @see WordPairs
     */
    public void saveAccount(View view) {

            if (isStoragePermissionGranted()) {
                //
                root = findExternalStorageRoot();
                rootPath = root.getAbsolutePath();
                //
                // naiveaac dir registration and csv copy from assets to dir naiveaac
                prepareTheSimsimDirectory();
                //
                realm= Realm.getDefaultInstance();
                //
                RealmResults<History> daCancellare = realm.where(History.class).findAll();
                realm.beginTransaction();
                daCancellare.deleteAllFromRealm();
                realm.commitTransaction();
                //
                File fI = AdvancedSettingsDataImportExportHelper.openFileInput (context,"images.csv");
                if (fI.exists()) {
                    Images.importFromCsv(context, realm);
                }
                File fV = AdvancedSettingsDataImportExportHelper.openFileInput (context,"videos.csv");
                if (fV.exists()) {
                    Videos.importFromCsv(context, realm);
                }
                //
                GameParameters.importFromCsv(context, realm);
                GrammaticalExceptions.importFromCsv(context, realm);
                ListsOfNames.importFromCsv(context, realm);
                Phrases.importFromCsv(context, realm);
                Stories.importFromCsv(context, realm);
                WordPairs.importFromCsv(context, realm);
                //
            }
            //
            // record default preferences
        sharedPref = this.getSharedPreferences(
                    getString(R.string.preference_file_key), Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(getString(R.string.preference_print_permissions), getString(R.string.character_n));
        editor.apply();
        editor.putString(getString(R.string.preference_list_mode), getString(R.string.sorted));
        editor.apply();
        editor.putInt("preference_AllowedMarginOfError", 20);
        editor.apply();
        // register the user in the shared preferences
        // and move on to the welcome activity
        EditText editText = (EditText) rootViewFragment.findViewById(R.id.editTextTextAccount);
        textPersonName = editText.getText().toString();
        //
        if (textPersonName.length()>0 && !filePath.equals(getString(R.string.non_trovato)))
        {
            // register the user
            registerPersonName(textPersonName);
            // register the linked image
            realm.beginTransaction();
            Images i = realm.createObject(Images.class);
            i.setDescrizione(textPersonName);
            i.setUri(filePath);
            realm.commitTransaction();
            realm.beginTransaction();
            Images iIo = realm.createObject(Images.class);
            iIo.setDescrizione(getString(R.string.io));
            iIo.setUri(filePath);
            realm.commitTransaction();
//
            Intent intent = new Intent(this,
                    ChoiseOfGameActivity.class);
            String message = (getString(R.string.puoi_accedere));
            intent.putExtra(EXTRA_MESSAGE, message);
            startActivity(intent);
        }
    }

    //
    /**
     * copy the csv files with initial settings and content such as images, videos and others from assets.
     *
     * @see #copyFileCsv
     * @see #copyAssets
     */
    public void prepareTheSimsimDirectory () {
        File f = new File(Environment.getExternalStorageDirectory(), getString(R.string.app_name));
        if (!f.exists()) {
            f.mkdirs();
            try {
                copyFileCsv("gameparameters.csv");
                copyFileCsv("grammaticalexceptions.csv");
                copyFileCsv("images.csv");
                copyFileCsv("listsofnames.csv");
                copyFileCsv("phrases.csv");
                copyFileCsv("pictogramsalltomodify.csv");
                copyFileCsv("stories.csv");
                copyFileCsv("videos.csv");
                copyFileCsv("wordpairs.csv");
            } catch (IOException e) {
                e.printStackTrace();
            }
            //
            copyAssets("images");
            copyAssets("videos");
            copyAssets("pdf");
        }
        //
    }
    //
    /**
     * copy the single csv file with initial settings and content from assets
     *
     * @param fileName string with the name of the file to be copied
     */
    public void copyFileCsv(String fileName) throws IOException {

        InputStream sourceStream = getAssets().open(getString(R.string.csv) + getString(R.string.character_slash) + fileName);
        InputStreamReader charSourceStream = new InputStreamReader(sourceStream, StandardCharsets.UTF_8);
        String destFileName = Environment.getExternalStorageDirectory()
                + getString(R.string.character_slash) + getString(R.string.app_name)
                + getString(R.string.character_slash) + fileName;
        FileOutputStream destStream = new FileOutputStream(destFileName);
        OutputStreamWriter charDestStream = new OutputStreamWriter(destStream, StandardCharsets.UTF_8);
        char[] buffer = new char[100];
        int charactersRead=0;
        while ( (charactersRead=charSourceStream.read(buffer))!=-1) {
            charDestStream.write(buffer,0,charactersRead);
        }
        charSourceStream.close();
        charDestStream.close();
        destStream.close();
        sourceStream.close();

    }
    //
    /**
     * check permissions.
     *
     * @return boolean whit true if permission is granted
     */
    public  boolean isStoragePermissionGranted() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    == PackageManager.PERMISSION_GRANTED) {
                // Log.v(TAGPERMISSION,getString(R.string.permission_is_granted));
                return true;
            } else {

                // Log.v(TAGPERMISSION,getString(R.string.permission_is_revoked));
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                return false;
            }
        }
        else { //permission is automatically granted on sdk<23 upon installation
            // Log.v(TAGPERMISSION,getString(R.string.permission_is_granted));
            return true;
        }
    }
    /**
     * on callback from SettingsFragment to this Activity
     *
     * @param v view root fragment view
     */
    @Override
    public void receiveResultSettings(View v) { rootViewFragment = v;  }
    //
    /**
     * copy assets images and videos to storage.
     * <p>
     * Refer to <a href="https://stackoverflow.com/questions/4447477/how-to-copy-files-from-assets-folder-to-sdcard">stackoverflow</a>
     * answer of <a href="https://stackoverflow.com/users/481239/rohith-nandakumar">Rohith Nandakumar</a>
     *
     * @param path string assets folder to copy
     * @see #copyFile(InputStream, OutputStream)
     */
    private void copyAssets(String path) {
        AssetManager assetManager = getAssets();
        String[] files = null;
        try {
            files = assetManager.list(path);
        } catch (IOException e) {
            // Log.e("tag", "Failed to get asset file list.", e);
        }
        if (files != null) {
            for (String filename : files) {
                InputStream in = null;
                OutputStream out = null;
                try {
                    in = assetManager.open(path + "/" + filename);
                    // qui
                    File outFile = new File(rootPath + "/" + getString(R.string.app_name), filename);
                    out = new FileOutputStream(outFile);
                    copyFile(in, out);
                } catch(IOException e) {
                    // Log.e("tag", "Failed to copy asset file: " + filename, e);
                }
                finally {
                    if (in != null) {
                        try {
                            in.close();
                        } catch (IOException e) {
                            // NOOP
                        }
                    }
                    if (out != null) {
                        try {
                            out.close();
                        } catch (IOException e) {
                            // NOOP
                        }
                    }
                }
            }
        }
    }
    /**
     * copy file.
     * <p>
     * Refer to <a href="https://stackoverflow.com/questions/4447477/how-to-copy-files-from-assets-folder-to-sdcard">stackoverflow</a>
     * answer of <a href="https://stackoverflow.com/users/481239/rohith-nandakumar">Rohith Nandakumar</a>
     *
     * @param in inputstream
     * @param out outputstream
     */
    private void copyFile(InputStream in, OutputStream out) throws IOException {
        byte[] buffer = new byte[1024];
        int read;
        while((read = in.read(buffer)) != -1){
            out.write(buffer, 0, read);
        }
    }



}

File: app/src/main/java/com/sampietro/NaiveAAC/activities/Account/AccountActivity.kt
package com.sampietro.NaiveAAC.activities.Account

import android.Manifest
import android.app.Activity
import android.app.Dialog
import com.sampietro.NaiveAAC.activities.BaseAndAbstractClass.AccountActivityAbstractClass
import android.os.Bundle
import com.sampietro.NaiveAAC.R
import com.sampietro.NaiveAAC.activities.Settings.AccountFragment
import com.sampietro.NaiveAAC.activities.Graphics.Sounds
import com.sampietro.NaiveAAC.activities.Graphics.Videos
import com.sampietro.NaiveAAC.activities.Game.GameParameters.GameParameters
import com.sampietro.NaiveAAC.activities.Grammar.GrammaticalExceptions
import com.sampietro.NaiveAAC.activities.Grammar.ListsOfNames
import com.sampietro.NaiveAAC.activities.Phrases.Phrases
import com.sampietro.NaiveAAC.activities.Stories.Stories
import com.sampietro.NaiveAAC.activities.WordPairs.WordPairs
import android.widget.EditText
import android.content.Intent
import com.sampietro.NaiveAAC.activities.Game.ChoiseOfGame.ChoiseOfGameActivity
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.ImageDecoder
import android.net.Uri
import android.view.ContextThemeWrapper
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.ImageButton
import android.widget.ImageView
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentManager
import com.google.android.material.snackbar.Snackbar
import com.sampietro.NaiveAAC.activities.Bluetooth.BluetoothDevices
//import com.sampietro.NaiveAAC.activities.DataStorage.DataStorageHelper
import com.sampietro.NaiveAAC.activities.DataStorage.DataStorageHelper.copyAssets
import com.sampietro.NaiveAAC.activities.DataStorage.DataStorageHelper.copyFileFromAssetsToInternalStorage
import com.sampietro.NaiveAAC.activities.DataStorage.DataStorageHelper.getFilePath
import com.sampietro.NaiveAAC.activities.Graphics.GraphicsAndPrintingHelper.showImage
//import com.sampietro.NaiveAAC.activities.Graphics.GraphicsAndPrintingHelper
import com.sampietro.NaiveAAC.activities.Graphics.Images
import com.sampietro.NaiveAAC.activities.history.History
import io.realm.Realm
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.lang.Exception
import java.net.URISyntaxException
import java.util.Objects

/**
 * <h1>AccountActivity</h1>
 *
 * **AccountActivity** create from assets the initial realm database , register the user
 * and move on to the welcome activity.
 * the initial realm database consists of
 *
 * 1) data from the Arasaac /pictograms/all API
 * 2) data from https://github.com/ian-hamlin/verb-data a collection of verbs and conjugations
 * 3) initial settings and content such as images, videos and others from assets
 *
 * @version     5.0, 01/04/2024
 * @see AccountActivityAbstractClass
 *
 * @see com.sampietro.NaiveAAC.activities.Settings.Utils.SettingsFragmentAbstractClass
 */
class AccountActivity : AccountActivityAbstractClass() {
    //
    var fragmentManager: FragmentManager? = null
    //
    var textPersonName: String? = null
    var textPassword: String? = null
    //
    /**
     * configurations of account start screen.
     *
     * @param savedInstanceState Define potentially saved parameters due to configurations changes.
     * @see setActivityResultLauncher
     *
     * @see AccountActionbarFragment
     *
     * @see AccountFragment
     *
     * @see android.app.Activity.onCreate
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_account)
        //
        realm = Realm.getDefaultInstance()
        //
        context = this
        //
        filePath = getString(R.string.non_trovato)
        //
        setActivityResultLauncher()
        //
        if (savedInstanceState == null) {
            fragmentManager = supportFragmentManager
            fragmentManager!!.beginTransaction()
                .add(AccountActionbarFragment(), "AccountActionbarFragment")
                .add(
                    R.id.settings_container,
                    AccountFragment(R.layout.activity_settings_account),
                    getString(R.string.account_fragment)
                )
                .commit()
        }
// Register the permissions callback, which handles the user's response to the
// system permissions dialog. Save the return value, an instance of
// ActivityResultLauncher. You can use either a val, as shown in this snippet,
// or a lateinit var in your onAttach() or onCreate() method.
        val requestPermissionLauncher =
            registerForActivityResult(
                ActivityResultContracts.RequestPermission()
            ) { isGranted: Boolean ->
                if (isGranted) {
                    // Permission is granted. Continue the action or workflow in your
                    // app.
                } else {
                    // Explain to the user that the feature is unavailable because the
                    // feature requires a permission that the user has denied. At the
                    // same time, respect the user's decision. Don't link to system
                    // settings in an effort to convince the user to change their
                    // decision.
                    // da COMPLETARE
                }
            }
        when {
            ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            ) == PackageManager.PERMISSION_GRANTED -> {
                // You can use the API that requires the permission.
            }
//            shouldShowRequestPermissionRationale(...) -> {
            // In an educational UI, explain to the user why your app requires this
            // permission for a specific feature to behave as expected, and what
            // features are disabled if it's declined. In this UI, include a
            // "cancel" or "no thanks" button that lets the user continue
            // using your app without granting the permission.
//            showInContextUI(...)
//        }
            else -> {
                // You can directly ask for the permission.
                // The registered ActivityResultCallback gets the result of this request.
                requestPermissionLauncher.launch(
                    Manifest.permission.WRITE_EXTERNAL_STORAGE)
            }
        }
        //
//        if (isStoragePermissionGranted) {
            //    Log.v(TAGPERMISSION,getString(R.string.permission_is_granted));
//        }
        //
    }
    /**
     * Called when the user taps the enter password button.
     *
     * @param v view of tapped button
     */
    fun enterPassword(v: View?) {
        //
        val d = Dialog(this)
        // Setting dialogview
        val window = d.window
        val wlp = window!!.attributes
        wlp.gravity = Gravity.BOTTOM
        wlp.flags = wlp.flags and WindowManager.LayoutParams.FLAG_DIM_BEHIND.inv()
        window.attributes = wlp
        //
        d.setTitle("Inserisci la password per accedere alle impostazioni \\n(opzionale se omessa , per accedere alle impostazioni, \\nverrà richiesto il risultato di un calcolo aritmetico)")
        d.setCancelable(false)
        d.setContentView(R.layout.activity_account_write_dialog)
        //
        val submitPasswordButton =
            d.findViewById<ImageButton>(R.id.submitPasswordButton)
        val cancelPasswordButton =
            d.findViewById<ImageButton>(R.id.cancelPasswordButton)
        val editTextPasswordAccount = d.findViewById<View>(R.id.editTextPasswordAccount) as EditText
        //
        editTextPasswordAccount.requestFocus()
        val imm = getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.showSoftInput(d.findViewById<View>(R.id.editTextPasswordAccount),
            InputMethodManager.SHOW_IMPLICIT
        )
        imm.hideSoftInputFromWindow(d.findViewById<View>(R.id.editTextPasswordAccount).getWindowToken(),
            InputMethodManager.HIDE_IMPLICIT_ONLY
        )
        //
        submitPasswordButton.setOnClickListener {
            //
            textPassword = editTextPasswordAccount.text.toString()
            imm.hideSoftInputFromWindow(editTextPasswordAccount.windowToken, 0)
            //
            d.cancel()
        }
        cancelPasswordButton.setOnClickListener {
            //
            imm.hideSoftInputFromWindow(editTextPasswordAccount.windowToken, 0)
            //
            d.cancel()
        }
        //
        d.show()
    }

    /**
     * Called when the user taps the save account button.
     *
     * if the case creates creates the initial realm database.
     *
     * @param view view of tapped button
     *
     * @see com.sampietro.NaiveAAC.activities.Settings.Utils.AdvancedSettingsDataImportExportHelper.findExternalStorageRoot
     * @see prepareTheSimsimDirectory
     *
     * @see registerPersonName
     * @see com.sampietro.NaiveAAC.activities.Settings.Utils.AdvancedSettingsDataImportExportHelper.openFileInput
     * @see Images
     *
     * @see Videos
     *
     * @see GameParameters
     *
     * @see GrammaticalExceptions
     *
     * @see ListsOfNames
     *
     * @see Phrases
     *
     * @see Stories
     *
     * @see WordPairs
     */
    fun saveAccount(view: View?) {
            // naiveaac dir registration and csv copy from assets to dir naiveaac
            prepareTheSimsimDirectory()
            //
            val daCancellare = realm.where(
                History::class.java
            ).findAll()
            realm.beginTransaction()
            daCancellare.deleteAllFromRealm()
            realm.commitTransaction()
            //
            Images.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            Sounds.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            Videos.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            //
            BluetoothDevices.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            GameParameters.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            GrammaticalExceptions.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            ListsOfNames.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            Phrases.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            Stories.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            WordPairs.importFromCsvFromInternalStorage(context, realm, getString(R.string.replace))
            //
        // record default preferences
        sharedPref = getSharedPreferences(
            getString(R.string.preference_file_key), MODE_PRIVATE
        )
        val editor = sharedPref.edit()
        editor.putString(
            getString(R.string.preference_print_permissions),
            getString(R.string.character_n)
        )
        editor.apply()
        editor.putString(getString(R.string.preference_title_writing_type), getString(R.string.uppercase))
        editor.apply()
        editor.putString(getString(R.string.preference_list_mode), getString(R.string.sorted))
        editor.apply()
        editor.putInt(getString(R.string.preference_allowed_margin_of_error), 20)
        editor.apply()
        // register the user in the shared preferences
        // and move on to the welcome activity
        val editText = findViewById<View>(R.id.editTextTextAccount) as EditText
        textPersonName = editText.text.toString()
        // default
        if (textPersonName!!.length <= 0) textPersonName = "utente"
        if (filePath == getString(R.string.non_trovato) || filePath == "da download") {
            val utenteFile = getFileStreamPath("utente.png")
            filePath = utenteFile.absolutePath
        }
        if (textPassword == null)   { textPassword = "nessuna password" }
                                    else
                                    { if (textPassword!!.length <= 0) textPassword = "nessuna password" }
        //
        if (textPersonName!!.length > 0 && filePath != getString(R.string.non_trovato)
            && filePath != "da download" && textPassword!!.length > 0
        ) {
            // register the user
            registerPersonName(textPersonName)
            // register the linked image
            realm.beginTransaction()
            val i = realm.createObject(
                Images::class.java
            )
            i.descrizione = textPersonName
            i.uri = filePath
            realm.commitTransaction()
            realm.beginTransaction()
            val iIo = realm.createObject(
                Images::class.java
            )
            iIo.descrizione = getString(R.string.io)
            iIo.uri = filePath
            realm.commitTransaction()
            // register the linked lists of names
            realm.beginTransaction()
            val listsOfNames = realm.createObject(ListsOfNames::class.java)
            // set the fields here
            listsOfNames.keyword = getString(R.string.famiglia)
            listsOfNames.word = textPersonName
            listsOfNames.elementActive = "A"
            listsOfNames.isMenuItem = "N"
            listsOfNames.fromAssets = ""
            realm.commitTransaction()
            // register the password
            registerPassword(textPassword)
            //
            val intent = Intent(
                this,
                ChoiseOfGameActivity::class.java
            )
            val message = getString(R.string.puoi_accedere)
            intent.putExtra(EXTRA_MESSAGE, message)
            startActivity(intent)
        }
    }
    //
    /**
     * copy the csv files with initial settings and content such as images, videos and others from assets.
     *
     * @see copyFileFromAssetsToInternalStorage
     *
     * @see copyAssets
     */
    fun prepareTheSimsimDirectory() {
        try {
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"bluetoothdevices.csv","bluetoothdevices.csv" )
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"gameparameters.csv","gameparameters.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"grammaticalexceptions.csv","grammaticalexceptions.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"images.csv","images.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"listsofnames.csv","listsofnames.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"phrases.csv","phrases.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"pictogramsalltomodify.csv","pictogramsalltomodify.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"sounds.csv","sounds.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"stories.csv","stories.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"videos.csv","videos.csv")
            copyFileFromAssetsToInternalStorage(context, getString(R.string.csv),"wordpairs.csv","wordpairs.csv")
        } catch (e: IOException) {
            e.printStackTrace()
        }
        //
        copyAssets(context,"images")
        copyAssets(context,"sounds")
        copyAssets(context,"videos")
        copyAssets(context,"pdf")
        //
    }
    // ActivityResultLauncher
    var imageSearchAccountActivityResultLauncher: ActivityResultLauncher<Intent>? = null
    //
    @JvmField
    var uri: Uri? = null
    //
    @JvmField
    var filePath: String? = null
    lateinit var byteArray: ByteArray
    //
    /**
     * setting callbacks to search for images and videos via ACTION_OPEN_DOCUMENT which is
     * the intent to choose a file via the system's file browser
     *
     *
     * Refer to [stackoverflow](https://stackoverflow.com/questions/62671106/onactivityresult-method-is-deprecated-what-is-the-alternative)
     * answer of [Muntashir Akon](https://stackoverflow.com/users/4147849/muntashir-akon)
     * and
     * Refer to [stackoverflow](https://stackoverflow.com/questions/56651444/deprecated-getbitmap-with-api-29-any-alternative-codes)
     * answer of [Ally](https://stackoverflow.com/users/6258197/ally)
     *
     * @see .getFilePath
     *
     * @see .showImage
     */
    fun setActivityResultLauncher() {
        imageSearchAccountActivityResultLauncher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult(),
            object : ActivityResultCallback<ActivityResult?> {
                //                @RequiresApi(Build.VERSION_CODES.P)
                override fun onActivityResult(result: ActivityResult?) {
                    if (result!!.resultCode == RESULT_OK) {
                        // There are no request codes
                        val resultData = result.data
                        // doSomeOperations();
                        uri = null
                        filePath = getString(R.string.non_trovato)
                        //
                        if (resultData != null) {
                            uri = Objects.requireNonNull(resultData).data
                            //
                            try {
                                filePath = getFilePath(context, uri)
                            } catch (e: URISyntaxException) {
                                e.printStackTrace()
                            }
                            //
                            if (filePath != getString(R.string.non_trovato))
                            {
                                if (filePath == "da download") {
                                    val ctw = ContextThemeWrapper(context, R.style.CustomSnackbarTheme)
                                    val snackbar = Snackbar.make(
                                        ctw,
                                        findViewById(R.id.imageviewaccounticon),
                                        "al momento l'app non è in grado di accedere ad immagini nella cartella download",
                                        10000
                                    )
                                    snackbar.setTextMaxLines(5)
                                    snackbar.setTextColor(Color.BLACK)
                                    snackbar.show()
                                }
                                else {
                                    val takeFlags =
                                        resultData.flags and Intent.FLAG_GRANT_READ_URI_PERMISSION
                                    context.contentResolver.takePersistableUriPermission(
                                        uri!!,
                                        takeFlags
                                    )
                                    //
                                    var bitmap: Bitmap? = null
                                    //
                                    try {
                                        val source = ImageDecoder.createSource(
                                            context.contentResolver,
                                            uri!!
                                        )
                                        bitmap = ImageDecoder.decodeBitmap(source) { decoder, _, _ ->
                                            decoder.setTargetSampleSize(1) // shrinking by
                                            decoder.isMutableRequired = true // this resolve the hardware type of bitmap problem
                                        }
                                    } catch (e: Exception) {
                                        e.printStackTrace()
                                    }
                                    //
                                    //
                                    val stream = ByteArrayOutputStream()
                                    bitmap!!.compress(Bitmap.CompressFormat.PNG, 100, stream)
                                    byteArray = stream.toByteArray()
                                    //
                                    val myImage: ImageView
                                    myImage = findViewById<View>(R.id.imageviewaccounticon) as ImageView
                                    showImage(context, uri, myImage)
                                }
                            }
                        }
                    }
                }
            })
        //
    }
    /**
     * Called when the user taps the image search button.
     *
     * @param v view of tapped button
     */
    fun imageSearch(v: View) {
        // ACTION_OPEN_DOCUMENT is the intent to choose a file via the system's file
        // browser.
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
        // Filter to only show results that can be "opened", such as a
        // file (as opposed to a list of contacts or timezones)
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        // Filter to show only images, using the image MIME data type.
        // If one wanted to search for ogg vorbis files, the type would be "audio/ogg".
        // To search for all documents available via installed storage providers,
        // it would be "*/*".
        intent.type = "image/*"
        /*  the instructions of the button */
        imageSearchAccountActivityResultLauncher!!.launch(intent)
    }
    companion object {
        //
        const val EXTRA_MESSAGE = "helloworldandroidMessage"
    }
}